### PR TITLE
Cleanup of token retrieval code - make storage token required, removing unneeded asserts

### PR DESCRIPTION
**Pull request by vladsud**
We do not allow or have any code path where we can make ODSP REST call without token - that would not work.
Make it obvious in the code and stop throwing unneeded asserts.

### CODE REVIEW

File: packages/drivers/odsp-driver/src/getUrlAndHeadersWithAuth.ts [hunk 1]
```diff
@@ -2,16 +2,15 @@
...
 	url: string,
 	// eslint-disable-next-line @rushstack/no-new-null
-	token: string | null,
+	token: string,
 	forceAccessTokenViaAuthorizationHeader: boolean,
 ): { url: string; headers: { [index: string]: string } } {
-	if (!token || token.length === 0) {
-		return { url, headers: {} };
-	}
+	assert(token.length > 0, "should be token");
```
Comment (pragya91) [inline packages/drivers/odsp-driver/src/getUrlAndHeadersWithAuth.ts:13]: I checked the usage of `getUrlAndHeadersWithAuth` in the code and this looks like a safe change, except there are a couple of places where we log telemetry as `headers:false` if authorization header was not present in the older code. I wonder if it makes sense to remove those logs as well. Don't have a strong opinion about this, not sure what was the original intent there.

File: packages/drivers/odsp-driver/src/odspUtils.ts [hunk 2]
```diff
@@ -369,12 +370,27 @@ export function evalBlobsAndTrees(snapshot: IOdspSnapshot): {
 	return { numTrees, numBlobs, encodedBlobsSize, decodedBlobsSize };
 }
 
+export function toInstrumentedOdspStorageTokenFetcher(
```
[the remark this answers is not in the document]
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspUtils.ts:373]: To double check, I run this query:

union Office_Fluid_FluidRuntime_*
| summarize count() by Data_headers

Result:

| headers | count |
| -------- | ------- |
|  | 9797673442 |
| true | 580302944 |

In other words, zero cases where it's false - if event has headers, it's always true

[the remark this answers is not in the document]
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspUtils.ts:373]: Actually, that does not tell us anything - we never log false :)

[the remark this answers is not in the document]
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspUtils.ts:373]: This is slightly better - first I found all events that could have headers, and then looked only at them:

union Office_Fluid_FluidRuntime_*
| where Data_headers == true
| summarize count() by Data_eventName

union Office_Fluid_FluidRuntime_*
| where Data_eventName in (
"fluid:telemetry:OdspDriver:uploadSummary_end",
"fluid:telemetry:OdspDriver:OpsFetch_end",
"fluid:telemetry:OdspDriver:TreesLatest_end",
"fluid:telemetry:OdspDriver:readDataBlob_cancel",
"fluid:telemetry:OdspDriver:readDataBlob_end",
"fluid:telemetry:OdspDriver:CreateNewFile_end",
"fluid:telemetry:OdspDriver:uploadSummary_cancel",
"fluid:telemetry:OdspDriver:OpsFetch",
"fluid:telemetry:OdspDriver:createNewEmptyFile_end",
"fluid:telemetry:OdspDriver:createNewFile_end",
"fluid:telemetry:OdspDriver:DeltaStorageOpsFetch",
"fluid:telemetry:OdspDriver:getVersions_cancel")
| summarize count() by Data_headers

| headers | count |
| -------- | ------- |
| true | 580363871 |
|  | 10 |

[the remark this answers is not in the document]
Comment (vladsud) [reply packages/drivers/odsp-driver/src/odspUtils.ts:373]: Ops, sorry, it was for other of your comments :)

**Review by pragya91 — APPROVED** [review]

**Pull request merged**